### PR TITLE
Product Sharing: UI for generating share message with AI

### PR DESCRIPTION
**Pull request by itsmeichigo**
Part of #9867 

## Description

This PR adds UI for the AI-generated sharing message for products feature following the latest update in pe5sF9-1AH-p2#comment-2256. Changes include:
- `ProductSharingMessageGenerationView` and view model: the UI for the new sharing sheet. 
- `FirstProductCreatedHostingController`, `ProductFormViewController`, `ProductsViewController`: persist `shareProductCoordinator` locally so that the coordinator is not deallocated after the containing methods are done executing.
- `ShareProductCoordinator`: Added a bottom sheet containing the new share sheet. 

## Testing instructions

Prerequisite: Make sure that you have access to a live WPCom site.
- On the Products tab, swipe to share a product or tap to share from its detail screen.
- Notice that a bottom sheet is displayed for AI-generated messages.
- Tap the button for generating message, notice that the loading state is displayed.
- When a message is generated, it should be displayed in a text editor. You should be able to change the content of the message.
- Tap Regenerate, a new message should be generated. Note: if you regenerate within 30 seconds of the last request, you'll get cached result from the last request.
- Tap Share, notice that the sheet is dismissed, and a native share sheet is displayed, and you can share both the product link and the generated message.

## Screenshots

https://github.com/woocommerce/woocommerce-ios/assets/5533851/117e5d9a-d58c-4068-8592-1df79969b58b

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationView.swift [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+                .renderedIf(viewModel.generationInProgress)
+
+            // Error message
+            viewModel.errorMessage.map { message in
+                Text(message).errorStyle()
+            }
+
+            // Action button to generate message
+            Button(action: {
+                Task { @MainActor in
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationView.swift]: Nit: Do we need to mention `@MainActor` here considering that `generateShareMessage` is already marked as `MainActor`?

File: WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationView.swift [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+                    .renderedIf(viewModel.messageContent.isEmpty)
+            }
+            .renderedIf(viewModel.generationInProgress == false)
+
+            // Skeleton view for loading state
+            Text(Constants.dummyText)
+                .secondaryBodyStyle()
+                .padding(Constants.placeholderInsets)
+                .background(RoundedRectangle(cornerRadius: Constants.cornerRadius).stroke(Color(uiColor: .secondarySystemFill)))
+                .redacted(reason: .placeholder)
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationView.swift]: Suggestion: Consider using `shimmering` effect along with this. Or do you think it is redundant due to the "Generating" spinner animation? 

https://github.com/woocommerce/woocommerce-ios/assets/524475/73d8faae-ea53-42ac-9a95-b2f81440004a

File: WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationViewModel.swift [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        errorMessage = nil
+        generationInProgress = true
+        do {
+            self.messageContent = try await withCheckedThrowingContinuation { continuation in
+                stores.dispatch(ProductAction.generateProductSharingMessage(siteID: siteID,
+                                                                            url: url,
+                                                                            languageCode: Locale.current.identifier,
+                                                                            completion: { result in
+                    continuation.resume(with: result)
+                }))
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationViewModel.swift]: Super nit/Suggestion: Consider extracting this to a private async function. Just to keep the continuation dealing code out for readability.

File: WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationViewModel.swift [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        self.viewTitle = String.localizedStringWithFormat(Localization.title, productName)
+    }
+
+    @MainActor
+    func generateShareMessage() async {
+        // TODO: Analytics
+        errorMessage = nil
+        generationInProgress = true
+        do {
+            self.messageContent = try await withCheckedThrowingContinuation { continuation in
```
Comment (selanthiraiyan) [inline WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationViewModel.swift]: Nit: I think we can skip the `self` here.

**Review by selanthiraiyan — APPROVED** [review]
Nice work. 🎉 

Everything works as expected except for the already logged [iPad popover issue from the first product created screen](https://github.com/woocommerce/woocommerce-ios/issues/9882). 

I left a few non-blocking suggestions.  🚀

File: WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationView.swift (hunk 1 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationView.swift]: Good point, removed in 8dad57feb40234ba3c130178da7b909431ce8ed2.

File: WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationView.swift (hunk 2 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationView.swift]: Nice, I didn't know about the shimmering effect! Added in de8e78e8dc262f229baa5db9eb6f761dd732df88.

File: WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationViewModel.swift (hunk 3 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationViewModel.swift]: 👍  Updated in 15fdd450a78f685d0e5f0f0dc60355659f238917.

File: WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationViewModel.swift (hunk 4 above)
Comment (itsmeichigo) [reply WooCommerce/Classes/ViewRelated/Products/AI/ProductSharingMessageGenerationViewModel.swift]: Also done in 15fdd450a78f685d0e5f0f0dc60355659f238917.

**Pull request merged**